src/main/resources/META-INF/rewrite/micronaut3-to-4.yml

```diff
@@ -0,0 +1,376 @@
+#
+# Copyright 2023 the original author or authors.
+# <p>
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# <p>
+# https://www.apache.org/licenses/LICENSE-2.0
+# <p>
+# <p>
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.micronaut.Micronaut3to4Migration
+displayName: Migrate from Micronaut 3.x to 4.x
+description: This recipe will apply changes required for migrating from Micronaut 3 to Micronaut 4.
+recipeList:
+  - org.openrewrite.java.micronaut.UpdateBuildToJava17
+  - org.openrewrite.gradle.UpdateGradleWrapper:
+      version: 8.x
+  - org.openrewrite.java.micronaut.UpdateBuildToMicronaut4Version
+  - org.openrewrite.java.micronaut.UpdateMicronautPlatformBom
+  - org.openrewrite.java.micronaut.UpdateBuildPlugins
+  - org.openrewrite.java.micronaut.AddSnakeYamlDependencyIfNeeded
+  - org.openrewrite.java.micronaut.UpdateJakartaAnnotations
+  - org.openrewrite.java.micronaut.UpdateJakartaPersistence
+  - org.openrewrite.java.micronaut.UpdateMicronautValidation
+  - org.openrewrite.java.micronaut.UpdateMicronautEmail
+  - org.openrewrite.java.micronaut.AddMicronautWebsocketDependencyIfNeeded
+  - org.openrewrite.java.micronaut.UpdateMicronautSession
+  - org.openrewrite.java.micronaut.AddMicronautRetryDependencyIfNeeded
+  - org.openrewrite.java.micronaut.UpdateMicronautSecurity
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.micronaut.UpdateBuildToJava17
+displayName: Update the Gradle or Maven build to a Java 17 baseline
+description: This recipe will update the Java version to 17 for a Gradle or Maven build.
+recipeList:
+  - org.openrewrite.gradle.UpdateJavaCompatibility:
+      version: 17
+  - org.openrewrite.maven.ChangePropertyValue:
+      key: jdk.version
+      newValue: 17
+  - org.openrewrite.maven.ChangePropertyValue:
+      key: release.version
+      newValue: 17
```

Comment:
We could pull these up into https://github.com/openrewrite/rewrite-migrate-java if you think that'd make sense. Then it could just be an include here rather than explicit handling. In rerwite-migrate-java would then also be the logical place to prevent downgrades of Java versions for instance, and additional changes such as Gradle toolchains and updating CI pipelines.

Reply:
Yes, I think it probably would make sense.